pkg/controller/controllerutil/controllerutil.go

```diff
@@ -304,7 +304,7 @@ func CreateOrPatch(ctx context.Context, c client.Client, obj client.Object, f Mu
 		result = OperationResultUpdated
 	}
 
-	if (hasBeforeStatus || hasAfterStatus) && !reflect.DeepEqual(beforeStatus, afterStatus) {
+	if hasBeforeStatus && hasAfterStatus && !reflect.DeepEqual(beforeStatus, afterStatus) {
```

Comment:
Wouldn't this check now fail if the object didn't have status before the call, but it does have it afterwards?

Reply:
You're right. The 2nd case would yield unexpected result:
```
hasBeforeStatus	F	T/F	T
hasAfterStatus	F	F/T	T
not eq			F* 	T**	T/F
--------------------
result OR			F	T	T/F
result AND		F	F	T/F

*cannot be True
**cannot be False
```

Reply:
OK there is nothing to improve here, since ``(A || B) && ...`` would stop processing all the rightmost, if either A or B yielded True...